lib/clients/AliClient.py

```diff
@@ -69,4 +154,397 @@ def _download_from_blobstore(self, blob_to_download_name, blob_download_target_p
+            if len(volume_list) == 0:
+                message = 'Volume with id {} is not found'.format(
+                volume_id)
+                raise Exception(message)
+            return volume_list[0]['Device']
+        except Exception as error:
+            self.logger.error(
+                '[ALI] ERROR: Unable to find device for attached volume {}.{}'.format(
+                    volume_id, error))
+            return None
+
+    def _create_attachment(self, volume_id, instance_id):
+        log_prefix = '[ATTACHMENT] [CREATE]'
+        attachment = None
+        device = None
+        attachment_creation_operation = None
+        try:
+            attachment_req_params = {
+                'InstanceId' : instance_id,
+                'DiskId' : volume_id
+            }
+            attachment_request = self._get_common_request('AttachDisk', attachment_req_params)
+            attachment_creation_operation = self.compute_client.do_action_with_exception(attachment_request)
+            
+            self._wait('Waiting for volume {} to get ready...'.format(volume_id),
+                       (lambda volume_id: self._is_volume_ready(volume_id, True)),
+                       None,
+                       volume_id)
+            
+            # Raise exception if device returned in None,
+            # as it might mean that disk was not attached properly
+            device = self._find_volume_device(volume_id)
+            if device:
+                self.logger.info(
+                    'Attached volume-id={}, device={}'.format(volume_id, device))
+                self._add_volume_device(volume_id, device)
+            else:
+                message = '{} ERROR: Device returned for volume-id={} is None'.format(
+                    log_prefix, volume_id)
+                raise Exception(message)
+
+            attachment = Attachment(0, volume_id, instance_id)
+            self._add_attachment(volume_id, instance_id)
+            self.logger.info(
+                '{} SUCCESS: volume-id={}, instance-id={}'.format(
+                    log_prefix, volume_id, instance_id))
+        except Exception as error:
+            message = '{} ERROR: volume-id={}, instance-id={}\n{}'.format(
+                log_prefix, volume_id, instance_id, error)
+            self.logger.error(message)
+
+            # The following lines are a workaround in case of inconsistency:
+            # The attachment process may end with throwing an Exception,
+            # but the attachment has been successful. Therefore, we must
+            # check whether the volume is attached and if yes, trigger the detachment
+            if attachment_creation_operation:
+                if device:
+                    self.logger.warning('[VOLUME] [DELETE] Volume is attached although the attaching process failed, '
+                                    'triggering detachment')
+                self.delete_attachment(volume_id, instance_id)
```

Comment:
I think indentation here is no correct, please check once.

Reply:
We trigger detachment based on `attachment_creation_operation` alone